ext/datadog_cov/datadog_cov.c

```diff
@@ -74,113 +63,108 @@ static VALUE dd_cov_allocate(VALUE klass)
 {
   struct dd_cov_data *dd_cov_data;
   VALUE obj = TypedData_Make_Struct(klass, struct dd_cov_data, &dd_cov_data_type, dd_cov_data);
+
   dd_cov_data->coverage = rb_hash_new();
-  dd_cov_data->root = Qnil;
-  dd_cov_data->ignored_path = Qnil;
-  dd_cov_data->mode = DD_COV_TARGET_FILES;
+  dd_cov_data->root = NULL;
+  dd_cov_data->root_len = 0;
+  dd_cov_data->ignored_path = NULL;
+  dd_cov_data->ignored_path_len = 0;
+  dd_cov_data->last_filename_ptr = 0;
+  dd_cov_data->frame_buffer = xcalloc(PROFILE_FRAMES_BUFFER_SIZE, sizeof(VALUE));
```

Comment:
Minor: Since we're only using a single memory position, it occurred to me a possible simplification would be to have a single `VALUE top_frame` in `dd_cov_data` and use that as a buffer instead. So this would be one less piece of dynamic memory to manage.

(This being C, you don't really need to pass in "an array" to rb_profile_frames, just a pointer to somewhere that can be written to -- and that includes a single field)